### PR TITLE
Lessen mount lock contention in layer store

### PR DESCRIPTION
**Pull request by kolyshkin**
While testing massive parallel container creation and removal, we see there's a lock contention in layer store that affects container creation performance (in other words, many threads are waiting for the lock). With this patches, container creation time (wall clock time) is about 2x to 3x less than before (measured with both aufs and overlay2 graph drivers).

The (global, per-layerstore) lock in question is held for the duration of
* CreateRWLayer()
* GetRWLayer()
* ReleaseRWLayer()
* CreateRWLayerByGraphID()

**The lock contention observed mostly between two instances of CreateRWLayer()**

Here's a stupid script I am using to test the performance before and after:
```bash
#!/bin/bash

NUMCTS=1000
PARALLEL=100

create() {
	# -v 'vol_{}:/vol'
	time ( seq $NUMCTS | parallel -j$PARALLEL docker create alpine top > /dev/null )
	echo "^^^ CREATE TIME ^^^"
}

remove() {
	time ( docker ps -qa | shuf | tail -n $NUMCTS | parallel -j$PARALLEL docker rm -f '{}' > /dev/null )
	echo "^^^ REMOVE TIME ^^^"
}

# precreate $NUMCTS containers
create
echo

# create another $NUMCTS while removing $NUMCTS containers
create &
remove &
wait

# remove the rest of it
remove
```

Results are:
1. Container creation time (the first timing print by the script) is about 40s before, about 20s after this patchset.

2. Concurrent container creation/removal time (the second and the third timing printed by the script) are about 70s before, 40s after this patchset.

3. Container removal times (last timing printed by the script) stay the same at about 30s.

### CODE REVIEW

**Comment by kolyshkin** [comment]
My concern is locking after this patches might not be adequate; perhaps we need to add a per-id (rather than global) lock (using `pkg/locker`) on top of it (but I have yet to look into this).

**Comment by kolyshkin** [comment]
> perhaps we need to add a per-id (rather than global) lock (using pkg/locker) on top of it

added (not sure if it's needed though)

**Comment by kolyshkin** [comment]
@dmcgowan PTAL

**Comment by dmcgowan** [comment]
A finer grain lock is needed in this case. The mounted layer operates on a reference map so there does require locking to prevent alterations on that object in parallel. It maybe possible to just add that lock to `*mountedLayer`, assuming the global lock has already synchronized creation of the same names.

**Comment by kolyshkin** [comment]
> A finer grain lock is needed in this case. The mounted layer operates on a reference map so there does require locking to prevent alterations on that object in parallel. It maybe possible to just add that lock to *mountedLayer, assuming the global lock has already synchronized creation of the same names.

@xinfengliu if you have time to address the above comments? I'm currently away for DockerCon and don't have time :(

**Comment by xinfengliu** [comment]
@kolyshkin , yes `mountedLayer` need a lock. Sorry I'm on China public holidays (May 1~3) now , I will spend some time on it but I'm not sure I can make it by the end of this week.

**Comment by kolyshkin** [comment]
@xinfengliu no problem, I think I fixed it.

@dmcgowan PTAL at the last commit (I decided to not use `RWMutex` for the sake of simplicity). Do we need same for `roLayer.references` map?

**Comment by dmcgowan** [comment]
> assuming the global lock has already synchronized creation of the same names.

I don't see this part being done. Possibly setting a placeholder in the map and clear it if there is an error. In that case a second call for the same name would get the conflicting error. I agree locking for the entire save is unnecessary just to prevent simultaneous creation of the same name.

**Comment by kolyshkin** [comment]
>> assuming the global lock has already synchronized creation of the same names.

> I don't see this part being done. Possibly setting a placeholder in the map and clear it if there is an error. In that case a second call for the same name would get the conflicting error. I agree locking for the entire save is unnecessary just to prevent simultaneous creation of the same name.


@dmcgowan thanks for review! Sorry I missed that part entirely; I think I know what you mean now, please see the updated commit. I had to introduce the third state of `ls.mount[name]` when it's not empty but the value is nil, meaning the name being created. All the other places treat `nil` as non-existent, except for `CreateRWLayer()` which treats it as a name conflict. I'm not in love with this approach as we can have potential nil dereferences but it looks cleaner than yet another lock or map (such as map of names being created).

Oh, if you can think of any test cases that can be created, I'm happy to implement those.

File: layer/layer_store.go [hunk 1]
```diff
@@ -491,13 +491,20 @@ func (ls *layerStore) CreateRWLayer(name string, parent ChainID, opts *CreateRWL
 	}
 
 	ls.mountL.Lock()
-	defer ls.mountL.Unlock()
-	m, ok := ls.mounts[name]
-	if ok {
+	if _, ok := ls.mounts[name]; ok {
 		return nil, ErrMountNameConflict
```
Comment (kolyshkin) [inline layer/layer_store.go]: I forgot to call Unlock on this path :-\ fix is coming

**Comment by kolyshkin** [comment]
Some changes based on @dmcgowan and my own review:
 - added locking to loadMount() (just for consistency/uniformity, techinically it's not required in there);
 - added missing Unlock() to an error path in CreateRWLayer;
 - simplified return statements in CreateRWLayer as we use named return values;
 - simplified map element presence check by dropping `ok` assignment and check (`ok && m != nil` --> `m != nil`).

**Review by dmcgowan — APPROVED** [review]
LGTM

File: layer/layer_store.go (hunk 1 above)
Comment (kolyshkin) [reply layer/layer_store.go]: fixed now
>+		ls.mountL.Unlock()
￼

**Comment by kolyshkin** [comment]
A comment from @tonistiigi :

> have some concerns about https://github.com/moby/moby/pull/39135. eg.in https://github.com/moby/moby/pull/39135/files#diff-5137c87d28c11ce0691d2f12e124409eR587 deleteRef, hasRef + remove is not atomic anymore. so if you call GetRWLayer after hasReference you get a rwlayer but the storage gets deleted in another goroutine

is to be addressed in a followup

**Pull request merged**